blockchain/v2/scheduler.go

```diff
@@ -383,3 +469,107 @@ func (sc *schedule) numBlockInState(targetState blockState) uint32 {
+	err := sc.touchPeer(event.peerID, event.time)
+	if err != nil {
+		return scPeerError{peerID: event.peerID, reason: err}, nil
+	}
+
+	err = sc.markReceived(event.peerID, event.block.Height, event.size, event.time)
+	if err != nil {
+		return scPeerError{peerID: event.peerID, reason: err}, nil
+	}
+
+	return scBlockReceived{peerID: event.peerID, block: event.block}, nil
+}
+
+func (sc *scheduler) handleBlockProcessed(event pcBlockProcessed) (Event, error) {
+	err := sc.markProcessed(event.height)
+	if err != nil {
+		return scSchedulerFail{reason: err}, nil // this should be fatal
+	}
+
+	if sc.allBlocksProcessed() {
+		return scFinishedEv{}, nil
+	}
+
+	return noOp, nil
+}
+
+// XXX: unify types peerError
+func (sc *scheduler) handlePeerError(event peerError) (Event, error) {
+	err := sc.removePeer(event.peerID)
+	if err != nil {
+		return scSchedulerFail{reason: err}, nil // xxx: bit extreme
+	}
+
+	return noOp, nil
+}
+
+func (sc *scheduler) handleTryPrunePeer(event tryPrunePeer) (Event, error) {
+	prunablePeers := sc.prunablePeers(sc.peerTimeout, sc.minRecvRate, event.time)
+
+	if len(prunablePeers) > 0 {
+		peerID := prunablePeers[0]
+		err := sc.removePeer(peerID)
+		if err != nil {
+			return scSchedulerFail{reason: err}, nil // xxx: should be fatal
+		}
+		return scPeerPruned{peerID: peerID}, nil
+	}
+	return noOp, nil
+}
+
+func (sc *scheduler) handleTrySchedule(event trySchedule) (Event, error) {
+	pendingBlocks := sc.numBlockInState(blockStatePending)
+	receivedBlocks := sc.numBlockInState(blockStateReceived)
+	todo := math.Min(float64(sc.targetPending-pendingBlocks), float64(sc.targetReceived-receivedBlocks))
+	if todo > 0 {
+		height := sc.minHeight()
+
+		if sc.getStateAtHeight(height) == blockStateNew {
+			allPeers := sc.getPeersAtHeight(height)
+			bestPeer := sc.selectPeer(allPeers) // XXX: maybe this should return a p2p.ID
+			err := sc.markPending(bestPeer.peerID, height, event.time)
+			if err != nil {
+				return scSchedulerFail{reason: err}, nil // XXX: peerError might be more appropriate
+			}
+			return scBlockRequest{peerID: bestPeer.peerID, height: height}, nil
+		}
+	}
+	return noOp, nil
+}
+
+func (sc *scheduler) handleStatusResponse(event bcStatusResponse) (Event, error) {
+	err := sc.setPeerHeight(event.peerID, event.height)
+	if err != nil {
+		return scPeerError{peerID: event.peerID, reason: err}, nil
+	}
+
+	return noOp, nil
+}
+
+func (sc *scheduler) handle(event Event) (Event, error) {
+	switch event := event.(type) {
+	case bcStatusResponse:
+		nextEvent, err := sc.handleStatusResponse(event)
+		return nextEvent, err
+	case trySchedule:
+		nextEvent, err := sc.handleTrySchedule(event)
+		return nextEvent, err
+	case tryPrunePeer:
+		nextEvent, err := sc.handleTryPrunePeer(event)
+		return nextEvent, err
+	case peerError:
+		nextEvent, err := sc.handlePeerError(event)
+		return nextEvent, err
+	case tmpBlockResponse:
+		nextEvent, err := sc.handleBlockResponse(event)
+		return nextEvent, err
+	}
```

Comment:
Need to add case for `addNewPeer`, `pcBlockProcessed`, `pcBlockVerificationFailure`.
Also add `default` for unhandled events vs. handled with result:
`return noOp, nil`